### PR TITLE
Reworked JS DOM, Zip and Unit Testing strategies

### PR DESCRIPTION
**Pull request by Innos**
Reworked JS DOM, Zip and Unit Testing strategies to process all tests with one execution

### CODE REVIEW

File: Open Judge System/Data/OJS.Data/Migrations/DefaultMigrationConfiguration.cs [hunk 1]
```diff
@@ -109,12 +108,12 @@ protected void SeedCheckers(OjsDbContext context)
 
         protected void SeedSubmissionTypes(OjsDbContext context)
         {
-            foreach (var entity in context.SubmissionTypes)
-            {
-                context.SubmissionTypes.Remove(entity);
-            }
+            //foreach (var entity in context.SubmissionTypes)
```
Comment (KristianMariyanov) [inline Open Judge System/Data/OJS.Data/Migrations/DefaultMigrationConfiguration.cs]: Revert this change

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs [hunk 2]
```diff
@@ -1,7 +1,12 @@
...
 {
     using System;
+    using System.Collections.Generic;
     using System.IO;
+    using System.Linq;
+    using System.Text.RegularExpressions;
+
+    using OJS.Workers.Common;
 
     public class IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy : NodeJsPreprocessExecuteAndRunUnitTestsWithMochaExecutionStrategy
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs]: Rename to `NodeJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs [hunk 3]
```diff
@@ -92,19 +104,97 @@ public class IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy : NodeJs
...
+
+            var processExecutionResult = this.ExecuteNodeJsProcess(executionContext, executor, string.Empty, arguments);
+
+            var mochaResult = JsonExecutionResult.Parse(processExecutionResult.ReceivedOutput);
+            var currentTest = 0;
+
+            foreach (var test in executionContext.Tests)
+            {
+                var message = "yes";
+                if (mochaResult.Error != string.Empty)
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs]: IMO it's better to check with `!string.IsNullOrEmpty(mochaResult.Error)`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs [hunk 4]
```diff
@@ -92,19 +104,97 @@ public class IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy : NodeJs
...
+                {
+                    message = mochaResult.Error;
+                }
+                else if (mochaResult.TestsErrors[currentTest] != string.Empty)
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs]: IMO it's better to check with `!string.IsNullOrEmpty(mochaResult.TestsErrors[currentTest])`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JsonExecutionResult.cs [hunk 5]
```diff
@@ -22,17 +23,28 @@ public class JsonExecutionResult
...
             try
             {
                 jsonTestResult = JObject.Parse(result.Trim().Replace("/*", InvalidJsonReplace).Replace("*/", InvalidJsonReplace));
                 totalPasses = (int)jsonTestResult["stats"]["passes"];
                 totalTests = (int)jsonTestResult["stats"]["tests"];
-                passed = totalPasses == 1;
+                var testsJs = (JArray)jsonTestResult["tests"];
+                for (int i = 0; i < testsJs.Count; i++)
+                {
+                    JToken token = jsonTestResult["tests"][i]["err"]["message"];
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JsonExecutionResult.cs]: use `var`

Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JsonExecutionResult.cs]: Can we refactor this expression in multiple lines for improving readability?

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JsonExecutionResult.cs [hunk 6]
```diff
@@ -22,17 +23,28 @@ public class JsonExecutionResult
...
+                    if (token != null)
+                    {
+                        errors.Add((string)jsonTestResult["tests"][i]["err"]["message"]);
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/JsonExecutionResult.cs]: maybe `errors.Add((string)token)` ?

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs [hunk 7]
```diff
@@ -34,222 +43,239 @@ public class NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecution
...
+        protected virtual string BuildTests(IEnumerable<TestContext> tests, int numberOfUserTests)
+        {
+            // Swap the testInput every X amount of tests where X is the amount of User Tests
+            // We keep track of the number we must swap at with the variable testCountRoof
+            var testCountRoof = numberOfUserTests;
+            var problemTests = tests.ToList();
+            var testsCode = problemTests[0].Input;
+
+            // We set the state of the tested entity in a beforeEach hook to ensure the user doesnt manipulate the entity
+            testsCode += @"
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunCodeAgainstUnitTestsWithMochaExecutionStrategy.cs]: You can use string interpolation in all strings

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunUnitTestsWithMochaExecutionStrategy.cs [hunk 8]
```diff
@@ -91,13 +103,23 @@ public class NodeJsPreprocessExecuteAndRunUnitTestsWithMochaExecutionStrategy :
             foreach (var test in executionContext.Tests)
             {
                 var processExecutionResult = this.ExecuteNodeJsProcess(executionContext, executor, test.Input, arguments);
-
                 var mochaResult = JsonExecutionResult.Parse(processExecutionResult.ReceivedOutput);
+
+                var message = "yes";
+                if (mochaResult.Error != string.Empty)
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsPreprocessExecuteAndRunUnitTestsWithMochaExecutionStrategy.cs]: IMO it's better to check with `!string.IsNullOrEmpty(mochaResult.Error)`

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 9]
```diff
@@ -91,152 +92,71 @@ public class NodeJsZipPreprocessExecuteAndRunUnitTestsWithDomAndMochaExecutionSt
...
-            var result = testFunc.call({}," + this.TestFuncVariables.Replace("'", string.Empty) + @", userBundleCode);
-
             Object.keys(bgCoderConsole)
-                .forEach(function (prop) {
-                    console[prop] = bgCoderConsole[prop];
-                });
-            done();
-        }";
+                .forEach(function(prop) {
+                console[prop] = bgCoderConsole[prop];
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: tab =>

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs [hunk 10]
```diff
@@ -273,6 +193,30 @@ public override ExecutionResult Execute(ExecutionContext executionContext)
...
+            var testsCount = 1;
+            foreach (var test in tests)
+            {
+                var code = Regex.Replace(test.Input, "([\\\\\"'])", "\\$1");
+                code = Regex.Replace(code, "[\r\n\t]*", string.Empty);
+
+                testsCode += @"
+                it('Test" + testsCount++ + @"', function(done) {
+                    this.timeout(10000);
+            	    let content = '" + code + @"';
```
Comment (KristianMariyanov) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/NodeJsZipPreprocessExecuteAndRunUnitTestsWithDOMAndMochaExecutionStrategy.cs]: one space to the right

File: Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs (hunk 4 above)
Comment (Innos) [inline Open Judge System/Workers/OJS.Workers.ExecutionStrategies/IoJsPreprocessExecuteAndRunJsDomUnitTestsExecutionStrategy.cs]: fixed
